src/API.jl

```diff
@@ -1138,7 +1138,7 @@ function precompile(ctx::Context, pkgs::Vector{PackageSpec}; internal_call::Bool
     ]
     stale_cache = Dict{StaleCacheKey, Bool}()
     exts = Dict{Base.PkgId, String}() # ext -> parent
-    # make a flat map of each dep and its direct deps
+    # make a flat map of each dep and its deps
     depsmap = Dict{Base.PkgId, Vector{Base.PkgId}}()
```

Comment:
I think the algorithm you are looking for here is Tarjan's SCC, since you already have the list of pkg->deps. You could also do a post-order DFS traversal, but the slight advantage of Tarjan is that it not only detects a cycle, but also computes the minimum list of entities in it. It is not quite as strong as doing a full dominator analysis (since it cannot sort within a cycle).

I have seemed to find myself in need of this quite often lately haha, such as
https://github.com/JuliaLang/julia/pull/49664. That is my favorite one, because of all the extra processing that has to be done there to compute the edges, which shows the flexibility of the Tarjan SCC algorithm. Though my cleanest implementation might be https://github.com/JuliaLang/julia/blob/master/src/staticdata_utils.c#L985-L1050, since it has the least extra state.

Reply:
Thanks! I think it will take me time to figure that out, so I'll get this reverted on master